spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/sql/CloudSqlEnvironmentPostProcessor.java

```diff
@@ -0,0 +1,203 @@
+ * @author Mike Eltsufin
+ * @author Chengyuan Zhao
+ * @author Eddú Meléndez
+ */
+public class CloudSqlEnvironmentPostProcessor implements EnvironmentPostProcessor {
+	private final static String CLOUD_SQL_PROPERTIES_PREFIX = "spring.cloud.gcp.sql.";
+
+	private static final Log LOGGER =
+			LogFactory.getLog(CloudSqlEnvironmentPostProcessor.class);
+
+	@Override
+	public void postProcessEnvironment(ConfigurableEnvironment environment, SpringApplication application) {
+		DatabaseType databaseType = getEnabledDatabaseType(environment);
+
+		if (databaseType != null) {
+			if (LOGGER.isInfoEnabled()) {
+				LOGGER.info("post-processing Cloud SQL properties for + " + databaseType.name());
+			}
+
+			CloudSqlJdbcInfoProvider cloudSqlJdbcInfoProvider = buildCloudSqlJdbcInfoProvider(environment, databaseType);
+
+			// configure default JDBC driver and username as fallback values when not specified
+			Map<String, Object> fallbackMap = new HashMap<>();
+			fallbackMap.put("spring.datasource.username", "root");
+			fallbackMap.put("spring.datasource.driver-class-name", cloudSqlJdbcInfoProvider.getJdbcDriverClass());
+			environment.getPropertySources()
+					.addLast(new MapPropertySource("CLOUD_SQL_DATA_SOURCE_FALLBACK", fallbackMap));
+
+			// always set the spring.datasource.url property in the environment
+			Map<String, Object> primaryMap = new HashMap<>();
+			primaryMap.put("spring.datasource.url", cloudSqlJdbcInfoProvider.getJdbcUrl());
+			environment.getPropertySources()
+					.addFirst(new MapPropertySource("CLOUD_SQL_DATA_SOURCE_URL", primaryMap));
+
+			setCredentials(environment, application);
+
+			// support usage metrics
+			CoreSocketFactory.setApplicationName("spring-cloud-gcp-sql/"
+					+ this.getClass().getPackage().getImplementationVersion());
+		}
+	}
+
+	private DatabaseType getEnabledDatabaseType(ConfigurableEnvironment environment) {
+		if (Boolean.parseBoolean(getSqlProperty(environment, "enabled", "true"))
+				&& isOnClasspath("javax.sql.DataSource")
+				&& isOnClasspath("org.springframework.jdbc.datasource.embedded.EmbeddedDatabaseType")
+				&& isOnClasspath("com.google.cloud.sql.CredentialFactory")) {
+			if (isOnClasspath("com.google.cloud.sql.mysql.SocketFactory")
+					&& isOnClasspath("com.mysql.cj.jdbc.Driver")) {
+				return DatabaseType.MYSQL;
+			}
+			else if (isOnClasspath("com.google.cloud.sql.postgres.SocketFactory")
+					&& isOnClasspath("org.postgresql.Driver")) {
+				return DatabaseType.POSTGRESQL;
+			}
+		}
+		return null;
+	}
+
+	private boolean isOnClasspath(String className) {
+		try {
+			ClassUtils.forName(className, null);
+			return true;
+		}
+		catch (ClassNotFoundException ex) {
+			return false;
+		}
+	}
+
+	private CloudSqlJdbcInfoProvider buildCloudSqlJdbcInfoProvider(ConfigurableEnvironment environment, DatabaseType databaseType) {
+		CloudSqlJdbcInfoProvider cloudSqlJdbcInfoProvider = new DefaultCloudSqlJdbcInfoProvider(
+				getSqlProperty(environment, "database-name", null),
```

Comment:
Should we be documenting these properties? Or at least enumerating them in some enum or constant set?

Reply:
What's the benefit?

Reply:
If we have to answer the question of "did we document everything", it would be easier to look in one spot and not play hide-and-seek with string constants. But it's up to you.

Reply:
You're right. I added them.